File: Course-Registration-System-main/java-(iteration_1)/src/CourseRegistrationSystem.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class CourseRegistrationSystem {
    private ArrayList<Transcript> transcripts;
    private ArrayList<Student> students;
    private ArrayList<Course> courses;
    private ArrayList<Lecturer> lecturers;
    private ArrayList<Advisor> advisors;

    public CourseRegistrationSystem() {
        this.transcripts = new ArrayList<>();
        this.students = new ArrayList<>();
        this.courses = new ArrayList<>();
        this.lecturers = new ArrayList<>();
        this.advisors = new ArrayList<>();
    }

    // creates transcripts
    public void createTranscriptList() {
        String directoryPath = "Data/Transcripts";

        try {
            // List all files in the directory
            List<java.nio.file.Path> filePaths = Files.list(Paths.get(directoryPath)).collect(Collectors.toList());

            // Iterate through each file and read the transcript
            for (Path filePath : filePaths) {
                TranscriptReader transcriptReader = new TranscriptReader();
                Transcript transcript = transcriptReader.readTranscriptFromFile(filePath.toString());
                getTranscripts().add(transcript);
            }
        } catch (IOException e) {
            e.printStackTrace(); // Handle the exception according to your needs
        }

    }

    // creates students
    public void createStudentList() {
        String filePath = "Data/student.json";
        students = (ArrayList<Student>) StudentReader.readStudentsFromFile(filePath);
    }

    // assigns transcripts to students
    public void assignTranscripts() {
        for (int i = 0; i < students.size(); i++) {
            for (int j = 0; j < transcripts.size(); j++) {
                if (transcripts.get(j).getId() == students.get(i).getId()) {
                    students.get(i).setTranscript(transcripts.get(j));
                }
            }
        }
    }

    // creates courses
    public void createCourseList() {
        String coursePath = "Data/course.json";
        courses = CourseReader.readCoursesFromJson(coursePath);
    }

    // creates lecturers
    public void createLecturerList() {
        String filePath = "Data/lecturer.json";
        lecturers = LecturerReader.readLecturersFromJsonFile(filePath);
    }

    // assigns lecturers to courses
    public void assignLecturerToCourses() {
        getCourses().get(0).setLecturer(getLecturers().get(0));
        getCourses().get(1).setLecturer(getLecturers().get(1));
        getCourses().get(2).setLecturer(getLecturers().get(2));
        getCourses().get(3).setLecturer(getLecturers().get(3));
        getCourses().get(4).setLecturer(getLecturers().get(4));
        getCourses().get(5).setLecturer(getLecturers().get(5));
        getCourses().get(6).setLecturer(getLecturers().get(6));
        getCourses().get(7).setLecturer(getLecturers().get(7));
        getCourses().get(8).setLecturer(getLecturers().get(8));
        getCourses().get(9).setLecturer(getLecturers().get(9));
        getCourses().get(10).setLecturer(getLecturers().get(10));
        getCourses().get(11).setLecturer(getLecturers().get(1));
        getCourses().get(12).setLecturer(getLecturers().get(11));
        getCourses().get(13).setLecturer(getLecturers().get(8));
        getCourses().get(14).setLecturer(getLecturers().get(9));
        getCourses().get(15).setLecturer(getLecturers().get(5));
        getCourses().get(16).setLecturer(getLecturers().get(10));

    }

    // assigns prerequisites to courses
    public void assignPrerequsitiesToCourses() {
        getCourses().get(11).getPrerequisite().add(getCourses().get(0)); // Operating System
        getCourses().get(10).getPrerequisite().add(getCourses().get(0)); // Database
        getCourses().get(14).getPrerequisite().add(getCourses().get(9)); // Modelling
        getCourses().get(16).getPrerequisite().add(getCourses().get(10)); // Advanced Unix
        getCourses().get(18).getPrerequisite().add(getCourses().get(3)); // Financial Engineering
    }

    // creates advisors from lecturers
    public void createAdvisor() {
        getAdvisors().add(new Advisor(lecturers.get(0)));
        getAdvisors().add(new Advisor(lecturers.get(1)));
        getAdvisors().add(new Advisor(lecturers.get(2)));
        getAdvisors().add(new Advisor(lecturers.get(3)));
        getAdvisors().add(new Advisor(lecturers.get(4)));

    }

    // assigns advisors to students
    public void assignAdvisorToStudents() {
        for (int i = 0; i < students.size(); i++) {
            for (int j = 0; j < advisors.size(); j++) {
                if (students.get(i).getAdvisorID() == advisors.get(j).getId()) {
                    students.get(i).setAdvisor(advisors.get(j));
                    advisors.get(j).getSupervisedStudentList().add(students.get(i));
                }
            }
        }
    }

    // checks the user type and logins the system as student or advisor
    public void login() {
        Scanner input = new Scanner(System.in);
        String selection;

        while (true) {
            System.out.println("----- Welcome to Marmara University Course Registration System -----");
            System.out.println("Type [1] to log in as advisor");
            System.out.println("Type [2] to log in as student");
            System.out.println("Type [q] to quit");
            System.out.print("-> ");
            selection = input.nextLine();

            if (!selection.equals("1") && !selection.equals("2") && !selection.equals("q")) {
                continue;
            }

            if (selection.equals("q")) {
                System.exit(0);
            } else if (selection.equals("1")) {
                advisorAuthentication();
            } else if (selection.equals("2")) {
                studentAuthentication();
            }
        }

    }

    // checks students id and password for authentication
    public void studentAuthentication() {
        Scanner input = new Scanner(System.in);
        Scanner input2 = new Scanner(System.in);
        int id;
        String password;
        boolean loginStatus = false;
        int currentStudentIndex = 0;

        do {
            System.out.print("Please enter your id: ");
            id = input.nextInt();

            System.out.print("Please enter your password: ");
            password = input2.nextLine();

            for (int i = 0; i < students.size(); i++) {
                if (id == students.get(i).getId() && password.equals(students.get(i).getPassword())) {
                    currentStudentIndex = i;
                    loginStatus = true;
                    break;
                }
            }
            // if autentication is successfully enter the system as student
            if (loginStatus == true) {
                System.out.println("\nAuthentication successful\n");
                StudentController controller = new StudentController(students.get(currentStudentIndex), courses);
                controller.showOptions();

            } else {
                System.out.println("\nincorrect user ID or password\n");
            }
        } while (loginStatus == false);
    }

    // Asks user to what action he/she wants to do

    // checks advisors id and password for authentication
    public void advisorAuthentication() {
        Scanner input = new Scanner(System.in);
        Scanner input2 = new Scanner(System.in);
        int id;
        String password;
        boolean loginStatus = false;
        int currentAdvisorIndex = 0;

        do {
            System.out.println("Please enter your id: ");
            id = input.nextInt();
            System.out.println("Please enter your password: ");
            password = input2.nextLine();

            for (int i = 0; i < advisors.size(); i++) {
                if (id == advisors.get(i).getId() && password.equals(advisors.get(i).getPassword())) {
                    loginStatus = true;
                    currentAdvisorIndex = i;
                    break;
                }
            }
            // if authentication is successfully enter the system as advisor
            if (loginStatus == true) {
                System.out.println("Authentication successful");
                AdvisorController advisorController = new AdvisorController(advisors.get(currentAdvisorIndex), courses);
                advisorController.showOptions();
            } else {
                System.out.println("Authentication failed!");
            }
        } while (loginStatus == false);
    }

    public ArrayList<Student> getStudents() {
        return students;
    }

    public void setStudents(ArrayList<Student> students) {
        this.students = students;
    }

    public ArrayList<Transcript> getTranscripts() {
        return transcripts;
    }

    public void setTranscripts(ArrayList<Transcript> transcripts) {
        this.transcripts = transcripts;
    }

    public ArrayList<Course> getCourses() {
        return courses;
    }

    public void setCourses(ArrayList<Course> courses) {
        this.courses = courses;
    }

    public ArrayList<Lecturer> getLecturers() {
        return lecturers;
    }

    public void setLecturers(ArrayList<Lecturer> lecturers) {
        this.lecturers = lecturers;
    }

    public ArrayList<Advisor> getAdvisors() {
        return advisors;
    }

    public void setAdvisors(ArrayList<Advisor> advisors) {
        this.advisors = advisors;
    }
}

File: Course-Registration-System-main/python-(iteration_2)/src/CourseRegistrationSystem.py
import os
import json
from Transcript import Transcript
from Student import Student
from Course import Course
from Advisor import Advisor
from Lecturer import Lecturer
from CourseSection import CourseSection
from StudentController import StudentController
from AdvisorController import AdvisorController
from TranscriptReader import readTranscriptFromFile
from StudentReader import readStudentsFromFile
from LecturerReader import readLecturersFromJsonFile
import logging

class CourseRegistrationSystem:
   
    def __init__(self):
        self.__transcripts = []
        self.__students = []
        self.__courses = []
        self.__lecturers = []
        self.__advisors = []
        self.logger = logging.getLogger(__name__)
        logging.basicConfig(level=logging.INFO)

    def createTranscriptList(self):
        directoryPath =  "./iteration-3/data/Transcripts"

        try:
            # List all files in the directory
            file_paths = [os.path.join(directoryPath, file) for file in os.listdir(directoryPath)]

            # Iterate through each file and read the transcript
            for file_path in file_paths:
                transcript = readTranscriptFromFile(file_path)
                self.__transcripts.append(transcript)
        except Exception as e:
            print(f"An error occurred: {e}")  # Handle the exception according to your needs

    def createStudentList(self):
        file_path = "./iteration-3/data/student.json"
        try:
            self.__students = readStudentsFromFile(file_path)
        except Exception as e:
            print(f"An error occurred: {e}")  # Handle the exception according to your needs

    def assignTranscripts(self):
        for student in self.__students:
            for transcript in self.__transcripts:

                if transcript.getId() == student.getId():
                    student.setTranscript(transcript)

    def createCourseList(self):
        course_path = "./iteration-3/data/course.json"
        with open(course_path, 'r') as file:
            course_data = json.load(file)
            self.__courses = [Course(
                course['courseName'],
                course['courseCode'],
                course['courseSemester'],
                course['courseType'],
                course['credit'],
                [CourseSection(
                    course['courseName'],
                    course['courseCode'],
                    course['courseSemester'],
                    course['courseType'],
                    course['credit'],
                    None,
                    course['prerequisite'],
                    course['lecturer'],
                    section['courseSectionCode'],
                    section['classroom'],
                    section['capacity'],
                    section['noOfEnrolledStudent'],
                    section['day'],
                    section['hour']
                ) for section in course['courseSections']],
                course['prerequisite'],
                course['lecturer']
            ) for course in course_data]


    def createLecturerList(self):
        file_path = "./iteration-3/data/lecturer.json"
        with open(file_path, 'r') as file:
            lecturer_data = json.load(file)
            self.__lecturers = readLecturersFromJsonFile(file_path)
           
    def assignLecturerToCourses(self):
         self.__courses[0].setLecturer(self.__lecturers[0])
         self.__courses[1].setLecturer(self.__lecturers[1])
         self.__courses[2].setLecturer(self.__lecturers[2])
         self.__courses[3].setLecturer(self.__lecturers[3]) 
         self.__courses[4].setLecturer(self.__lecturers[4]) 
         self.__courses[5].setLecturer(self.__lecturers[5])
         self.__courses[6].setLecturer(self.__lecturers[6])
         self.__courses[7].setLecturer(self.__lecturers[7]) 
         self.__courses[8].setLecturer(self.__lecturers[8])
         self.__courses[9].setLecturer(self.__lecturers[9])
         self.__courses[10].setLecturer(self.__lecturers[10])
         self.__courses[11].setLecturer(self.__lecturers[1]) 
         self.__courses[12].setLecturer(self.__lecturers[11])
         self.__courses[13].setLecturer(self.__lecturers[8])
         self.__courses[14].setLecturer(self.__lecturers[9]) 
         self.__courses[15].setLecturer(self.__lecturers[5]) 
         self.__courses[16].setLecturer(self.__lecturers[10])


    def assignPrerequisitesToCourses(self):
        self.__courses[11].getPrerequisite().append(self.__courses[0]) # Operating System
        self.__courses[10].getPrerequisite().append(self.__courses[0]) # Database
        self.__courses[14].getPrerequisite().append(self.__courses[9]) # Modelling
        self.__courses[16].getPrerequisite().append(self.__courses[10]) # Advanced Unix
        self.__courses[18].getPrerequisite().append(self.__courses[3]) # Financial Engineering

    def createAdvisor(self):
        for lecturer in self.__lecturers[:5]:
            advisor = Advisor(lecturer)
            self.__advisors.append(advisor)

    def assignAdvisorToStudents(self):
        for student in self.__students:
            for advisor in self.__advisors:
                if student.getAdvisorId() == advisor.getId():
                    student.setAdvisor(advisor)
                    advisor.getSupervisedStudentList().append(student)


    def login(self):
        while True:
            print("----- Welcome to Marmara University Course Registration System -----")
            print("Type [1] to log in as advisor")
            print("Type [2] to log in as student")
            print("Type [q] to quit")
            selection = input("-> ")

            if selection not in ["1", "2", "q"]:
                continue

            if selection == "q":
                exit()
            elif selection == "1":
                self.advisorAuthentication()
            elif selection == "2":
                self.studentAuthentication()

    def studentAuthentication(self):
        while True:
            try:
                id = int(input("Please enter your id: "))
                password = input("Please enter your password: ")

                for i, student in enumerate(self.__students):
                    if id == student.getId() and password == student.getPassword():
                        self.logger.info("Authentication successful for student with ID: %d", id)
                        studentController = StudentController(student, self.__courses)
                        studentController.showOptions()
                        return
                self.logger.warning("Authentication failed for student with ID: %d", id)
                print("\nIncorrect user ID or password\n")
            except ValueError:
                self.logger.error("Invalid ID entered")
                print("\nPlease enter a valid ID\n")

    def advisorAuthentication(self):
        loginStatus = False
        currentAdvisorIndex = 0

        while not loginStatus:
            try:
                advisorId = int(input("Please enter your id: "))
                password = input("Please enter your password: ")

                for i, advisor in enumerate(self.__advisors):
                    if advisorId.__eq__(advisor.getId()) and password.__eq__(advisor.getPassword()) :
                        self.logger.info("Authentication successful for advisor with ID: %d", advisor.getId())
                        loginStatus = True
                        currentAdvisorIndex = i
                        break

                if loginStatus:
                    advisorController = AdvisorController(self.__advisors[currentAdvisorIndex], self.__courses)
                    advisorController.showOptions()
                else:
                    self.logger.warning("Authentication failed for advisor with ID: %d", advisorId)

            except ValueError:
                print("Please enter a valid ID")

    def getStudents(self):
        return self.__students

    def setStudents(self, students):
        self.__students = students

    def getTranscripts(self):
        return self.__transcripts

    def setTranscripts(self, transcripts):
        self.__transcripts = transcripts

    def getCourses(self):
        return self.__courses

    def setCourses(self, courses):
        self.__courses = courses

    def getLecturers(self):
        return self.__lecturers

    def setLecturers(self, lecturers):
        self.__lecturers = lecturers

    def getAdvisors(self):
        return self.__advisors

    def setAdvisors(self, advisors):
        self.__advisors = advisors
